### PR TITLE
[SDL-0293] Enable OEM exclusive apps support

### LINKED ISSUE #1876: [SDL 0293] Enable OEM exclusive apps support
**Proposal:** [Enable OEM exclusive apps support](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0293-vehicle-type-filter.md)

> This feature will enable SDL adopters to provide exclusive apps to their users depending on vehicle type. The proposal describes a way to share vehicle type information before sending the Register App Interface request.

**Review:** https://github.com/smartdevicelink/sdl_evolution/issues/971

**Steering Committee Decision:** 
The Steering Committee voted to accept this proposal with the following revisions:

The author is to update the proposal to include the requested requirement for the Java Suite Library to have two pull requests submitted for this feature. The reasoning as to why is documented in [this comment ](https://github.com/smartdevicelink/sdl_evolution/issues/971#issuecomment-730557977)and should be included in the proposal.

The proposal .md file was updated to reflect these revisions on 12/14/2020.

### PR DESCRIPTION
**Pull request by yoooriii**
Fixes #1876 

### Risk
This PR makes **minor** API changes.

### Testing Plan
- [x] I have verified that I have not introduced new warnings in this PR (or explain why below)
- [x] I have run the unit tests with this PR
- [x] I have tested this PR against Core and verified behavior (if applicable, if not applicable, explain why below).

#### Unit Tests
SDLControlFramePayloadRPCStartServiceAckSpec.m updated to cover new added vehicleType field.
Also it tests SDLVehicleType that comes in bson data.

#### Core Tests
[List of tests performed against Core and behaviors verified]

Core version / branch / commit hash / module tested against: [Core implementation](https://github.com/LuxoftSDL/sdl_core/tree/feature/sdl_0293_enable_oem_exclusive_apps_support)
HMI name / version / branch / commit hash / module tested against: [HMI  implementation](https://github.com/LuxoftSDL/sdl_hmi/tree/feature/sdl_0293_enable_oem_exclusive_apps_support)

### Summary
* Added new optional property **SDLVehicleType * vehicleType** to the struct **SDLControlFramePayloadRPCStartServiceAck**
* new protocol **SDLVehicleTypeHandler** added
* added 2 new methods to **SDLProtocolDelegate** `shouldProceedWithVehicleType` and `doDisconnectWithVehicleType`
* the protocol listeners logic updated

##### Breaking Changes
* no breaking changes

##### Enhancements
* no enhancements

##### Bug Fixes
* fixed memory leak in SDLControlFramePayloadRPCStartServiceAck.m (free bson data)

### Tasks Remaining:
- [x] implement unit tests

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

**Comment by yoooriii** [comment]
@SatbirTanda @standa1-ford

File: Example Apps/Example ObjC/ProxyManager.m [hunk 1]
```diff
@@ -318,6 +318,14 @@ - (nullable SDLLifecycleConfigurationUpdate *)managerShouldUpdateLifecycleToLang
     return update;
 }
 
+- (BOOL)didReceiveVehicleType:(SDLVehicleType *)type {
+    SDLLogD(@"The vehicle type is: %@", type);
+    return YES;
+    // check the type and return Yes or No depending on your business logic.
+    // usage example:
+    // return [type.make isEqualToString:@"Ford"];
```
Comment (standa1-ford) [inline Example Apps/Example ObjC/ProxyManager.m:326]: is this a TODO?

File: SmartDeviceLink/private/SDLControlFramePayloadConstants.m [hunk 2]
```diff
@@ -26,3 +26,9 @@
 char *const SDLControlFrameWidthKey = "width";
 char *const SDLControlFrameTCPIPAddressKey = "tcpIpAddress";
 char *const SDLControlFrameTCPPortKey = "tcpPort";
+char *const SDLControlFrameVehicleHardVersion = "systemHardwareVersion";
+char *const SDLControlFrameVehicleMake = "make";
+char *const SDLControlFrameVehicleModel = "model";
+char *const SDLControlFrameVehicleModelYear = "model year";
```
Comment (standa1-ford) [inline SmartDeviceLink/private/SDLControlFramePayloadConstants.m:32]:
```suggestion
char *const SDLControlFrameVehicleModelYear = "modelYear";
```

File: SmartDeviceLink/private/SDLControlFramePayloadRPCStartServiceAck.m [hunk 3]
```diff
@@ -23,9 +24,12 @@ @interface SDLControlFramePayloadRPCStartServiceAck ()
 @property (copy, nonatomic, readwrite, nullable) NSArray<NSString *> *secondaryTransports;
 @property (copy, nonatomic, readwrite, nullable) NSArray<NSNumber *> *audioServiceTransports;
 @property (copy, nonatomic, readwrite, nullable) NSArray<NSNumber *> *videoServiceTransports;
+@property (strong, nonatomic, readwrite, nullable) SDLVehicleType *vehicleType;
 
 @end
 
+SDLVehicleType *sdl_parseVehicleType(BsonObject *const payloadObject);
+BOOL sdl_putStringValue(BsonObject *const payloadObject, const char *key, NSString *value);
```
Comment (standa1-ford) [inline SmartDeviceLink/private/SDLControlFramePayloadRPCStartServiceAck.m:32]: this looks like a private function definition; if its a must they should be inside the interface declaration

File: SmartDeviceLink/private/SDLControlFramePayloadRPCStartServiceAck.m [hunk 4]
```diff
@@ -145,12 +160,48 @@ - (void)sdl_parse:(NSData *)data {
...
+            free(buf);
+        }
+    }
+    return success;
+}
+
+SDLVehicleType *sdl_parseVehicleType(BsonObject *const payloadObject) {
+    SDLVehicleType *vehicleType = nil;
+    NSString *make = payloadObject ? sdl_getStringValue(payloadObject, SDLControlFrameVehicleMake) : nil;
+    if (0 < make.length) {
```
Comment (standa1-ford) [inline SmartDeviceLink/private/SDLControlFramePayloadRPCStartServiceAck.m:194]:
```suggestion
    if (make.length > 0) {
```

File: SmartDeviceLink/private/SDLLifecycleManager.m [hunk 5]
```diff
@@ -394,7 +401,13 @@ - (void)didEnterStateConnected {
             return;
         }
 
-        weakSelf.registerResponse = (SDLRegisterAppInterfaceResponse *)response;
+        SDLRegisterAppInterfaceResponse *raiResponse = (SDLRegisterAppInterfaceResponse *)response;
+        const BOOL shouldProceed = (nil == raiResponse) || [weakSelf shouldProceedWithVehicleType:raiResponse.vehicleType];
```
Comment (standa1-ford) [inline SmartDeviceLink/private/SDLLifecycleManager.m:405]:
```suggestion
        const BOOL shouldProceed = (raiResponse == nil) || [weakSelf shouldProceedWithVehicleType:raiResponse.vehicleType];
```

File: SmartDeviceLinkTests/ProtocolSpecs/MessageSpecs/SDLControlFramePayloadConstantsSpec.m [hunk 6]
```diff
@@ -32,6 +32,18 @@
         expect(SDLControlFrameWidthKey).to(equal(@"width"));
         expect(SDLControlFrameTCPIPAddressKey).to(equal(@"tcpIpAddress"));
         expect(SDLControlFrameTCPPortKey).to(equal(@"tcpPort"));
+        expect(SDLControlFrameVehicleHardVersion).to(equal("systemHardwareVersion"));
+        expect(strcmp("systemHardwareVersion", SDLControlFrameVehicleHardVersion)).to(equal(0));
+        expect(SDLControlFrameVehicleMake).to(equal("make"));
+        expect(strcmp("make", SDLControlFrameVehicleMake)).to(equal(0));
+        expect(SDLControlFrameVehicleModel).to(equal("model"));
+        expect(strcmp("model", SDLControlFrameVehicleModel)).to(equal(0));
+        expect(SDLControlFrameVehicleModelYear).to(equal("model year"));
```
Comment (standa1-ford) [inline SmartDeviceLinkTests/ProtocolSpecs/MessageSpecs/SDLControlFramePayloadConstantsSpec.m:41]: After other fix:
```suggestion
        expect(SDLControlFrameVehicleModelYear).to(equal("modelYear"));
```

File: SmartDeviceLink/private/SDLControlFramePayloadConstants.h [hunk 7]
```diff
@@ -26,3 +26,10 @@ extern char *const SDLControlFrameVideoServiceTransportsKey;
 extern char *const SDLControlFrameTCPIPAddressKey;
 extern char *const SDLControlFrameTCPPortKey;
 extern char *const SDLControlFrameAuthTokenKey;
+// Vehicle type related keys
```
Comment (standa1-ford) [inline SmartDeviceLink/private/SDLControlFramePayloadConstants.h:29]:
```suggestion
```

File: SmartDeviceLink/private/SDLControlFramePayloadRPCStartServiceAck.m [hunk 8]
```diff
@@ -145,12 +160,48 @@ - (void)sdl_parse:(NSData *)data {
...
+NSString *sdl_getStringValue(BsonObject *const payloadObject, const char *key) {
+    const char *cValue = key && payloadObject ? bson_object_get_string(payloadObject, key) : NULL;
+    return cValue ? [NSString stringWithUTF8String:cValue] : nil;
+}
+
+BOOL sdl_putStringValue(BsonObject *const payloadObject, const char *key, NSString *value) {
+    BOOL success = NO;
+    if (payloadObject && key && value) {
+        // rationale: bson_object_put_string does not respect const so we have to make a tmp copy
+        const NSUInteger leng = 1 + [value lengthOfBytesUsingEncoding:NSUTF8StringEncoding];
```
Comment (standa1-ford) [inline SmartDeviceLink/private/SDLControlFramePayloadRPCStartServiceAck.m:180]:
```suggestion
        const NSUInteger leng = [value lengthOfBytesUsingEncoding:NSUTF8StringEncoding] + 1;
```

File: SmartDeviceLink/private/SDLLifecycleManager.m [hunk 9]
```diff
@@ -969,4 +982,32 @@ - (void)sdl_remoteHardwareDidUnregister:(SDLRPCNotificationNotification *)notifi
 
 @end
 
+#pragma mark - SDLVehicleTypeHandler
+
+@implementation SDLLifecycleManager (VehicleTypeHandler)
+
+- (BOOL)shouldProceedWithVehicleType:(SDLVehicleType *)vehicleType {
+    if ([self.enabledVehicleType isEqual:vehicleType]) {
+        // no need to check it twice (it has already been enabled previously)
```
Comment (standa1-ford) [inline SmartDeviceLink/private/SDLLifecycleManager.m:991]:
```suggestion
```

File: SmartDeviceLink/private/SDLLifecycleManager.m [hunk 10]
```diff
@@ -969,4 +982,32 @@ - (void)sdl_remoteHardwareDidUnregister:(SDLRPCNotificationNotification *)notifi
...
+        return YES;
+    }
+    // if the delegate present (not nil) and responds to the optional selector then ask it or answer YES instead.
```
Comment (standa1-ford) [inline SmartDeviceLink/private/SDLLifecycleManager.m:994]:
```suggestion
```

**Review by standa1-ford — APPROVED** [review]

**Comment by joeljfischer** [comment]
@standa1-ford, you can't review and give a bunch of feedback and simultaneously approve a PR. I'm considering this PR as _not_ approved by Ford given that you placed feedback.

**Comment by yoooriii** [comment]
Hello @standa1-ford. I updated the PR as it was requested. Could you please re-review it? Thank you.

**Review by standa1-ford — APPROVED** [review]
Approved. cc: @joeljfischer
Thanks, will keep that in mind!

**Comment by jordynmackool** [comment]
@yoooriii / @standa1-ford 

 Can you please resolve the merge conflicts on this PR prior to us reviewing?

**Comment by yoooriii** [comment]
Hello @jordynmackool. I have merged the develop branch into this one and resolved 2 conflicts in the xcode project file. Could you please proceed? Thank you.

**Comment by jordynmackool** [comment]
@yoooriii  the proposal markdown file has been updated per the revisions included in the accepted with revisions review issue: [SDL 0293 Revisions - Enable OEM exclusive apps support](https://github.com/smartdevicelink/sdl_evolution/issues/1108). 

Please make the needed updates to this PR and then tag @FrankElias77 to review when ready. Thanks!

**Comment by yoooriii** [comment]
Hello @FrankElias77. I have updated the PR as it was requested. Could you please review it?
Also please advice if I need to add more tests to increase the code coverage.
Thank you.

File: Example Apps/Example ObjC/ProxyManager.m [hunk 11]
```diff
@@ -318,6 +318,11 @@ - (nullable SDLLifecycleConfigurationUpdate *)managerShouldUpdateLifecycleToLang
     return update;
 }
 
+- (BOOL)didReceiveSystemInfo:(SDLSystemInfo *)systemInfo {
+    SDLLogD(@"The system is: %@", systemInfo);
+    return YES;
+}
+
```
Comment (FrankElias77) [inline Example Apps/Example ObjC/ProxyManager.m:325]: Why was this function added in this file?

**Review by FrankElias77 — CHANGES_REQUESTED** [review]
@yoooriii 
- Warning in Objective-c/Swift example apps
- Missing changes in SDLRegisterAppInterfaceResponse that are mentioned in proposal(section: "Addition of "RegisterAppInterface")
- Warnings in the spec files

**Comment by FrankElias77** [comment]
@yoooriii Please tag me once these changes are made. Thanks!

**Comment by joeljfischer** [comment]
@yoooriii, the Java Suite team has written [a new PR](https://github.com/smartdevicelink/sdl_java_suite/pull/1618) for this proposal. Please update this PR or open a new one to match.

**Comment by yoooriii** [comment]
Hello @FrankElias77. I have updated the PR, could you proceed please? Tnx.
Please note, since a few properties were declared as deprecated I had to update a few extra files and tests to avoid warnings.

File: SmartDeviceLink/public/SDLSystemInfo.m [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+    if (![object isKindOfClass:self.class]) {
+        return NO;
+    }
+    typeof(self) other = object;
+    // handle cases where both properties are nil
+    return ((self.vehicleType == other.vehicleType) || [self.vehicleType isEqual:other.vehicleType]) &&
+            ((self.systemSoftwareVersion == other.systemSoftwareVersion) || [self.systemSoftwareVersion isEqualToString:other.systemSoftwareVersion]) &&
+            ((self.systemHardwareVersion == other.systemHardwareVersion) || [self.systemHardwareVersion isEqualToString:other.systemHardwareVersion]);
+}
+
```
Comment (FrankElias77) [inline SmartDeviceLink/public/SDLSystemInfo.m:49]: Implement description for this class please
```suggestion

- (NSString *)description
{
    return [NSString stringWithFormat:@"<#format string#>", <#arguments#>];
}

```

**Review by FrankElias77 — CHANGES_REQUESTED** [review]
@yoooriii Tag me please, once these changes are done

- Increase SDLGlobals's MaxProtocolVersion to 5.4.0
- Warnings on runtime related to Umbrella header. You can fix that by Changing target membership for "SDLTransportType.h" from public to project

File: SmartDeviceLink/private/SDLProtocol.m [hunk 13]
```diff
@@ -507,12 +507,16 @@ - (void)sdl_processMessages {
...
 
+    SDLControlFramePayloadRPCStartServiceAck *startServiceACKPayload = nil;
+    SDLVersion *version = nil;
     // V5+ Packet
     if (startServiceACK.header.version >= 5) {
         switch (startServiceACK.header.serviceType) {
             case SDLServiceTypeRPC: {
-                SDLControlFramePayloadRPCStartServiceAck *startServiceACKPayload = [[SDLControlFramePayloadRPCStartServiceAck alloc] initWithData:startServiceACK.payload];
+                startServiceACKPayload = [[SDLControlFramePayloadRPCStartServiceAck alloc] initWithData:startServiceACK.payload];
+                version = startServiceACKPayload.protocolVersion == nil ? nil : [SDLVersion versionWithString:startServiceACKPayload.protocolVersion];
```
Comment (FrankElias77) [inline SmartDeviceLink/private/SDLProtocol.m:519]:
```suggestion
                version = startServiceACKPayload.protocolVersion == nil ? [SDLVersion versionWithMajor:startServiceACK.header.version minor:0 patch:0] : [SDLVersion versionWithString:startServiceACKPayload.protocolVersion];
```

File: SmartDeviceLink/private/SDLProtocol.m [hunk 14]
```diff
@@ -521,22 +525,24 @@ - (void)protocol:(SDLProtocol *)protocol didReceiveStartServiceACK:(SDLProtocolM
                     self.hashId = startServiceACKPayload.hashId;
                 }
 
-                [SDLGlobals sharedGlobals].maxHeadUnitProtocolVersion = (startServiceACKPayload.protocolVersion != nil) ? [SDLVersion versionWithString:startServiceACKPayload.protocolVersion] : [SDLVersion versionWithMajor:startServiceACK.header.version minor:0 patch:0];
+                [SDLGlobals sharedGlobals].maxHeadUnitProtocolVersion = (version != nil) ? version : [SDLVersion versionWithMajor:startServiceACK.header.version minor:0 patch:0];
```
Comment (FrankElias77) [inline SmartDeviceLink/private/SDLProtocol.m:528]:
```suggestion
                [SDLGlobals sharedGlobals].maxHeadUnitProtocolVersion = version;
```

**Comment by yoooriii** [comment]
Hello @FrankElias77. I have updated the PR. Could you please review? Thank you.

**Review by FrankElias77 — CHANGES_REQUESTED** [review]
@yoooriii Added one more small change. Please tag me once the changes are done!

**Comment by yoooriii** [comment]
Hello @FrankElias77. I have fixed the PR. Could you please proceed? Thank you.

**Review by FrankElias77 — APPROVED** [review]

**Comment by joeljfischer** [comment]
Closing in favor of #1919. This pull request had a good amount of unnecessary changes, contained a number of bugs and style issues, and was missing a good chunk of unit tests. Due to the sdl_ios project coming to a development cycle end very soon it was determined that the project maintainer could no longer continue back and forth reviews on this PR. The PM has created a new PR that includes a simpler solution and additional unit tests. Thank you for your understanding.

**Pull request closed without merging**